File: HW/HW12/Main.py
import heapq

def minCostToSupplyWater(self, n: int, wells: List[int], pipes: List[List[int]]) -> int:
    graph = collections.defaultdict(list)
    for i in range(n):
        graph[0].append((wells[i], i + 1))  # Connect virtual node to houses
    for house1, house2, cost in pipes:
        graph[house1].append((cost, house2))
        graph[house2].append((cost, house1))

    visited = set()
    totalCost = 0
    pq = [(0, 0)]  # Priority queue: (cost, house)

    while len(visited) < n + 1:
        cost, house = heapq.heappop(pq)
        if house in visited:
            continue
        visited.add(house)
        totalCost += cost

        for next_cost, neighbor in graph[house]:
            heapq.heappush(pq, (next_cost, neighbor))

    return totalCost

File: HW/HW12/Main.java
public int minCostToSupplyWater(int n, int[] wells, int[][] pipes) {
    // Create a list to store all edges (both wells and pipes)
    List<int[]> edges = new ArrayList<>();
    for (int i = 0; i < n; i++) {
        edges.add(new int[]{0, i + 1, wells[i]}); // Connect virtual node '0' to houses
    }
    edges.addAll(Arrays.asList(pipes));

    // Sort edges by cost (ascending)
    Collections.sort(edges, Comparator.comparingInt(a -> a[2]));

    // Union-Find for tracking connected components
    UnionFind uf = new UnionFind(n + 1); // Include the virtual node

    int totalCost = 0;
    for (int[] edge : edges) {
        int house1 = edge[0], house2 = edge[1], cost = edge[2];
        if (uf.union(house1, house2)) {
            totalCost += cost;
        }
    }
    return totalCost;
}
class UnionFind {
    private int[] root; // Stores the root of each component (house) 
    private int[] rank; // Stores the rank (tree height) of each component

    public UnionFind(int size) {
        root = new int[size];
        rank = new int[size];
        // Initialize each house as its own component
        for (int i = 0; i < size; i++) {
            root[i] = i;
            rank[i] = 1; 
        }
    }

    // Find the root of a component (house)
    public int find(int x) {
        if (x == root[x]) {
            return x;
        }
        // Path compression optimization: Update the root as we traverse
        return root[x] = find(root[x]);
    }

    // Union two components by merging the trees
    public boolean union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        if (rootX != rootY) {
            // Merge based on rank for optimization
            if (rank[rootX] > rank[rootY]) {
                root[rootY] = rootX;
            } else if (rank[rootX] < rank[rootY]) {
                root[rootX] = rootY;
            } else {
                root[rootY] = rootX;
                rank[rootX] += 1;
            }
            return true; // Successfully merged
        } else {
            return false; // Already belong to the same component
        }
    }
}
